src/main/java/io/pravega/connectors/flink/FlinkPravegaReader.java

```diff
@@ -220,6 +258,150 @@ private void triggerCheckpoint(String checkpointIdentifier) throws FlinkExceptio
+        }
+    }
+
+    /**
+     * Gauge for getting the scope name information from reader group.
+     */
+    private static class ScopeNameGauge implements Gauge<String> {
+
+        private final ReaderGroup readerGroup;
+
+        public ScopeNameGauge(ReaderGroup readerGroup) {
+            this.readerGroup = readerGroup;
+        }
+
+        @Override
+        public String getValue() {
+            return readerGroup.getScope();
+        }
+    }
+
+    /**
+     * Gauge for getting online readers information from reader group.
+     */
+    private static class OnlineReadersGauge implements Gauge<String> {
+
+        private final ReaderGroup readerGroup;
+        private final String separator = ",";
+
+        public OnlineReadersGauge(ReaderGroup readerGroup) {
+            this.readerGroup = readerGroup;
+        }
+
+        @Override
+        public String getValue() {
+            return readerGroup.getOnlineReaders().stream().collect(Collectors.joining(separator));
+        }
+    }
+
+    /**
+     * Gauge for getting stream name information from reader group.
+     */
+    private static class StreamNamesGauge implements Gauge<String> {
+
+        private final ReaderGroup readerGroup;
+        private final String separator = ",";
+
+        public StreamNamesGauge(ReaderGroup readerGroup) {
+            this.readerGroup = readerGroup;
+        }
+
+        @Override
+        public String getValue() {
+            return readerGroup.getStreamNames().stream().collect(Collectors.joining(separator));
+        }
+    }
+
+    /**
+     * Gauge for getting position information of each segment of a stream from reader group.
+     */
+    private static class SegmentPositionsGauge implements Gauge<String> {
+
+        private final ReaderGroup readerGroup;
+        private final String stream;
+
+        public SegmentPositionsGauge(ReaderGroup readerGroup, String stream) {
+            this.readerGroup = readerGroup;
+            this.stream = stream;
+        }
+
+        @Override
+        public String getValue() {
+            StringBuilder builder = new StringBuilder();
+            builder.append("stream=").append(stream).append(", segments={");
+            Map<Stream, StreamCut> streamCuts = readerGroup.getStreamCuts();
+            Optional<Map.Entry<Stream, StreamCut>> optionalStreamCutEntry =
+                    streamCuts.entrySet().stream().filter(e -> e.getKey().getStreamName().equals(stream)).findFirst();
+            if (optionalStreamCutEntry.isPresent()) {
+                optionalStreamCutEntry.get().getValue().asImpl().getPositions().entrySet().stream().forEach(entry -> {
+                    builder.append(entry.getKey()).append("=").append(entry.getValue()).append(", ");
+                });
+            }
+            builder.append("}");
+            return builder.toString();
+        }
+    }
+
+    /**
+     * register reader group metrics
+     *
+     */
+    private void registerMetrics() {
```

Comment:
I am unsure about the threading model for Flink metrics and for the reader group object.   Obviously the metrics are pulled using a different thread than the `run` thread.   Anything to be concerned about?

Reply:
From what I have tested so far, I did not see any issues in metrics reporting by Flink.